File: leetcode/javascript/hashing/validAnagram(easy)/validAnagram.js
/* https://leetcode.com/problems/valid-anagram/description/

Учитывая две строки s и t, верните true, если t является анаграммой s, и false в противном случае.
Анаграмма — это слово или фраза, образованная путем перестановки букв другого слова или фразы, обычно с использованием
всех исходных букв ровно один раз.

Input: s = "anagram", t = "nagaram"
Output: true

Input: s = "rat", t = "car"
Output: false
*/

/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isAnagram = function(s, t) {
    if (s.length !== t.length) return false;
    let map = new Map();

    for (let char of s) {
        if (map.has(char)) map.set(char, map.get(char) + 1);
        else map.set(char, 1);
    }

    for (let char of t) {
        if (!map.has(char) || map.get(char) <= 0) return false;
        map.set(char, map.get(char) - 1);
    }

    return true;
};

console.log(isAnagram("anagram", "nagaram")); // true

File: leetcode/go/hashing/validAnagram(easy)/validAnagram.go
package main

import "fmt"

/* 242. Valid Anagram
https://leetcode.com/problems/valid-anagram/description/

Учитывая две строки s и t, верните true, если t является анаграммой s, и false в противном случае.
Анаграмма — это слово или фраза, образованная путем перестановки букв другого слова или фразы, обычно с использованием
всех исходных букв ровно один раз.

Input: s = "anagram", t = "nagaram"
Output: true

Input: s = "rat", t = "car"
Output: false
*/

func main() {
	fmt.Println(isAnagram("anagram", "nagaram")) // true
}

// isAnagram проверяет, являются ли строки анаграммами.
// time: O(n), space: O(n)
func isAnagram(s string, t string) bool {
	// Если длины строк разные, они не могут быть анаграммами
	if len(s) != len(t) {
		return false
	}

	// Создаем map для подсчета символов
	charCount := make(map[rune]int)

	// Подсчитываем количество каждого символа в первой строке
	for _, char := range s {
		charCount[char]++
	}

	// Проверяем вторую строку
	for _, char := range t {
		// Если символа нет в первой строке или его недостаточно
		if count, exists := charCount[char]; !exists || count <= 0 {
			return false
		}
		// Уменьшаем счетчик символа
		charCount[char]--
	}

	return true // Все символы в обеих строках присутствуют одинаковое количество раз
}
